lib/uniffi-zcash-cli/src/diff/grep_item.rs

```diff
@@ -0,0 +1,252 @@
+use anyhow::anyhow;
+use clap::ColorChoice;
+use nu_ansi_term::{self, Color};
+use public_api::tokens::Token;
+use serde::Deserialize;
+
+use super::grep_type::GrepType;
+
+const TOO_MANY_GREP_MATCHES_NUM: usize = 30;
+
+#[derive(Deserialize)]
+struct Range {
+    start: Position,
+    end: Position,
+}
+
+#[derive(Deserialize)]
+struct Position {
+    line: usize,
+    column: usize,
+}
+
+#[derive(Deserialize)]
+struct Match {
+    range: Range,
+    file: String,
+    lines: String,
+}
+
+// An item representing the thing to search(grep) for in the code.
+#[derive(Default)]
+pub(crate) struct GrepItem {
+    pub(crate) api_diff: String,
+    pub(crate) color: ColorChoice,
+    pattern: String,
+    grep_type: GrepType,
+    grep_output: Option<String>,
+    matches: Vec<Match>,
+}
+
+impl GrepItem {
+    /// Runs ast-grep (sg command), which contextually searches for a pattern, based on the programming
+    /// language
+    pub(crate) fn grep(&mut self, grep_dir: &String) -> anyhow::Result<&Self> {
+        let output = std::process::Command::new("sg")
+            .arg("run")
+            .arg("-p")
+            .arg(&self.pattern)
+            .arg("-l")
+            .arg("rs")
+            .arg("--color")
+            .arg(self.color.to_string())
+            .arg("--heading")
+            .arg("always")
+            .arg("--context")
+            .arg("1")
+            .arg(grep_dir)
+            .arg("--json")
+            .output();
+
+        match output {
```

Comment:
remove the match, handle error directly on setting the output and unnest the part in `Ok(o) => { ... }`
```rs
let output = ...
    .with_context(|e| format!("bla bla"))?;`
```